Переработан WebDriverManager.

evgenykuzin opened the pull request:
новый класс BySet для хранения тегов, классов, и тд по которым будем кликать.

Commit:
Переработан WebDriverManager.
новый класс BySet для хранения тегов, классов, и тд по которым будем кликать.

Commit:
в BySet изменен тип аргумента на 'String...'.
фикс WebDriverManager.

alozhkin commented on src/main/java/selenium/BySet.java:
```diff
@@ -0,0 +1,71 @@
+
+public class BySet extends HashSet<By> {
+
+    public class Builder {
```
А зачем билдер? Он разве на что-то влияет?

evgenykuzin replied:
Подумал, что так будет проще добавлять другие селекторы через точечку. Можно, конечно и по-другому реализовать.

alozhkin replied:
Ну раз он ничего не делает, просто убери его. Его методы спокойно будут работать в BySet

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,93 @@
+    private Html currentHtml;
+
+    public WebDriverManager() {
+        init("");
```
Можно просто вызвать WebDriverManager(new Link("")), а код init вставить в конструктор с Link

evgenykuzin replied:
Хочется оставить конструктор без аргументов.

alozhkin replied:
Я имел ввиду вот так: https://imgur.com/a/RDmYuho

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,93 @@
+        for (By by : bySet) {
+            elements.addAll(driver.findElements(by));
+        }
+        BlockingQueue<Html> htmls = new ArrayBlockingQueue<>(1000);
```
Почему BlockingQueue?

evgenykuzin replied:
Ну.. это рудимент от прошлого решения. Что посоветуешь на замену?

alozhkin replied:
Просто лист. Если тебе нужно ограничение по размеру, то if (size > 1000) break;

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    public WebDriverManager(String url) {
```
https://github.com/alozhkin/OPD-2020/pull/8#discussion_r392664004

evgenykuzin replied:
Ок. Чистый так чистый.

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.concurrent.BlockingQueue;
+
+public class WebDriverManager {
+    private String url;
```
Лучше использовать Link и переименовать в currentUrl

evgenykuzin replied:
Согласен

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public Html getHtml() {
+        return new Html(driver.getPageSource(), new Link(url));
```
А почему просто не вернуть currentHtml?

evgenykuzin replied:
потому что смысл метода в возвращении текущего html. Есть смысл переименовать метод в `parseHtml()`

alozhkin replied:
Не правильно прочитал, всё ок.

alozhkin reviewed (changes requested):
Посмотрел ваш код, выглядит неплохо. Но запустить и проверить у себя как он работает я не могу.

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,93 @@
+        }
+        BlockingQueue<Html> htmls = new ArrayBlockingQueue<>(1000);
+        for (WebElement element : elements) {
+            try {
```
Кстати пропустил. Нельзя кидать Exception в цикле, это очень медленно.

Commit:
Фиксы

Commit:
Фиксы #2

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -1,7 +1,6 @@
 package selenium;
```
#XX имеет особый смысл на github, это ссылка на pull request или issue. В будущем используй №.

Commit:
ignore this

Commit:
Merge branch 'develop' of https://github.com/alozhkin/OPD-2020 into develop

alozhkin commented on src/test/java/selenium/WebDriverManagerTest.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Test
+    public void testGetDynamicContent() {
+        String chromeDriverPath = "C:\\Users\\JekaJops\\IntelliJIDEAProjects\\words_extractor\\src\\main\\resources\\drivers\\chromedriver1.exe";
```
В конфиге создай файл local.properties и напиши project.path=... Всё, можешь доставать из System. В удалённом репозитории не должно быть ничего связанного с локальной файловой системой. 
\+ В качестве разделителя следует использовать /, потому что он работает и в Windows и в Linux.

alozhkin commented on src/main/java/selenium/BySet.java:
```diff
@@ -0,0 +1,72 @@
+
+    public class Builder {
+
+        public Builder addTagNames(String... selectors) {
```
Ты можешь просто возращать BySet и делать вызовы в цепочку без всякого билдера.

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,82 @@
+        for (By by : bySet) {
+            elements.addAll(driver.findElements(by));
+        }
+        ArrayList<Html> htmls = new ArrayList<>();
```
program to interface, not implementation

Commit:
refactoring.
new methods (you can click or not click to get dynamic content).
problem: connecting to some sites is too slow.

Commit:
Merge branch 'develop' of https://github.com/alozhkin/OPD-2020 into develop

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    private static WebDriver initDriver() {
```
Добавь сюда исправления от Маши #16

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    public Html parseHtml() {
+        return new Html(driver.getPageSource(), new Link(driver.getCurrentUrl()));
```
А тут currentLink и спарсенный сайт не могут разойтись в результате редиректа?

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,139 @@
+        return currentLink;
+    }
+
+    public void resetCurrentHtml() {
```
reload скорее

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,139 @@
+        Collection<String> result = new ArrayList<>();
+        List<WebElement> elements = new ArrayList<>();
+        int i = 0;
+        int border = 1000; // �� ������������� �������. ������ ����� ������ ����� ���������� ������ �� ��������.
```
k0дNpOвÖ4Кä

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,139 @@
+                elements.addAll(driver.findElements(by));
+            }
+            if (elements.isEmpty()) {
+                connect(currentLink);
```
Это перезагрузка типа?

alozhkin commented on src/main/java/selenium/WebDriverManager.java:
```diff
@@ -0,0 +1,139 @@
+            }
+
+        }
+
```
Многовато пустых строк

alozhkin reviewed (commented):
Норм, появилось пара вопросов о том, что делает код

Commit:
добавил логер и опции для драйвера

alozhkin commented on src/test/java/selenium/WebDriverManagerTest.java:
```diff
@@ -0,0 +1,45 @@
+        webDriverManager.connect(link);
+        Html dynamicHtml = webDriverManager.getCurrentHtml();
+        Assertions.assertNotEquals(parsedHtml, dynamicHtml);
+//        Collection<String> words = driverManager.getNewWords(parsedHtml, dynamicHtml);
```
Лучше избегать закоменченного кода.
Мартин, гл 17, c5
